webui/src/App.js

```diff
@@ -21,107 +21,187 @@ import {
+  }, [isWsReady, wsRef]);
 
-  ws.send(JSON.stringify(msg));
-}
+  wsCallbacksRef.current.values = function(msg) {
+    const webUIData = webUIDataRef.current;
+    if (webUIData.curValues.length < MAX_SIZE) {
+      webUIData.curValues.push(msg.values);
+    } else {
+      webUIData.curValues[webUIData.oldest] = msg.values;
+      webUIData.oldest = (webUIData.oldest + 1) % MAX_SIZE;
+    }
+  };
 
-wsCallbacks.values = function(msg) {
-  if (kCurValues.length < max_size) {
-    kCurValues.push(msg.values);
-  } else {
-    kCurValues[oldest] = msg.values;
-    oldest = (oldest + 1) % max_size;
-  }
-};
+  wsCallbacksRef.current.thresholds = function(msg) {
+    // Modify thresholds array in place instead of replacing it so that animation loops can have a stable reference.
+    webUIDataRef.current.curThresholds.length = 0;
+    webUIDataRef.current.curThresholds.push(...msg.thresholds);
+  };
+
+  useEffect(() => {
+    let cleaningUp = false;
+    const webUIData = webUIDataRef.current;
 
-wsCallbacks.thresholds = function(msg) {
-  kCurThresholds = msg.thresholds;
-};
+    if (!defaults) {
+      // If defaults are loading or reloading, don't connect.
+      return;
+    }
+
+    // Ensure values history reset and default thresholds are set.
+    webUIData.curValues.length = 0;
+    webUIData.curValues.push(new Array(defaults.thresholds.length).fill(0));
+    webUIData.oldest = 0;
+    webUIDataRef.current.curThresholds.length = 0;
+    webUIDataRef.current.curThresholds.push(...defaults.thresholds);
 
-connect();
+    const ws = new WebSocket('ws://' + window.location.host + '/ws');
+    wsRef.current = ws;
+
+    ws.addEventListener('open', function(ev) {
+      setIsWsReady(true);
+    });
+    
+    ws.addEventListener('error', function(ev) {
+      ws.close();
+    });
+
+    ws.addEventListener('close', function(ev) {
+      if (!cleaningUp) {
+        onCloseWs();
+      }
+    });
+
+    ws.addEventListener('message', function(ev) {
+      const data = JSON.parse(ev.data)
+      const action = data[0];
+      const msg = data[1];
+
+      if (wsCallbacksRef.current[action]) {
+        wsCallbacksRef.current[action](msg);
+      }
+    });
+
+    return () => {
+      cleaningUp = true;
+      setIsWsReady(false);
+      ws.close();
+    };
+  }, [defaults, onCloseWs]);
+
+  return { emit, isWsReady, webUIDataRef, wsCallbacksRef };
+}
 
 // An interactive display of the current values obtained by the backend.
 // Also has functionality to manipulate thresholds.
 function ValueMonitor(props) {
-  const index = parseInt(props.index)
+  const { emit, index, webUIDataRef } = props;
   const thresholdLabelRef = React.useRef(null);
   const valueLabelRef = React.useRef(null);
   const canvasRef = React.useRef(null);
+  const curValues = webUIDataRef.current.curValues;
+  const curThresholds = webUIDataRef.current.curThresholds;
 
-  function EmitValue(val) {
+  const EmitValue = useCallback((val) => {
```

Comment:
Just for my understanding since I'm not well versed in React, what does `useCallback` here do better instead of defining a function like I had previously?

Reply:
Every render, the ValueMonitor component function runs and EmitValue is created as a new function. Since EmitValue is a dependency of the useEffect call, a new function every render causes the effect to rerun every render. useCallback remembers the function passed to it on the first render and keeps returning that so you have a function with a stable reference (until its own dependencies change).

Reply:
I'm not sure how clear that explanation was but the point is to avoid an infinite render loop.

Reply:
Ahhhh that makes sense! So if `EmitValue` wasn't part of the dependency list, it would be okay to not use `useCallback` as it wouldn't trigger the recursive loop?

Reply:
Right! Actually I was thinking about it and I'm not sure if it actually causes a render loop in this case, but it defeats the purpose of the dependency list. The effect only reruns when its dependencies change, but if you include a dependency that is different every render, it's effectively the same as omitting the dependency list.